framework/auth/views.py

```diff
@@ -378,30 +378,58 @@ def auth_register(auth):
 
     raise HTTPError(http.BAD_REQUEST)
 
-
-def auth_logout(redirect_url=None):
+@collect_auth
+def auth_logout(auth, redirect_url=None, next_url=None):
     """
     Log out, delete current session and remove OSF cookie.
-    Redirect to CAS logout which clears sessions and cookies for CAS and Shibboleth (if any).
-    Final landing page may vary.
+    If next url is valid and auth is logged in, redirect to CAS logout endpoint with the current request url as service.
+    If next url is valid and auth is logged out, redirect directly to the next url.
+    Otherwise, redirect to CAS logout or login endpoint with redirect url as service.
+    The CAS logout endpoint which clears sessions and cookies for CAS and Shibboleth.
     HTTP Method: GET
 
-    :param redirect_url: url to redirect user after CAS logout, default is 'goodbye'
-    :return:
+    Note 1: OSF tells CAS where it wants to be redirected back after successful logout. However, CAS logout flow may not
+    respect this url if user is authenticated through remote identity provider.
+    Note 2: The name of the query parameter is `next`, `next_url` is used to avoid python reserved word.
+
+    :param auth: the authentication context
+    :param redirect_url: url to DIRECTLY redirect after CAS logout, default is `OSF/goodbye`
+    :param next_url: url to redirect after OSF logout, which is after CAS logout
+    :return: the response
     """
 
-    # OSF tells CAS where it wants to be redirected back after successful logout.
-    # However, CAS logout flow may not respect this url if user is authenticated through remote identity provider.
-    redirect_url = redirect_url or request.args.get('redirect_url') or web_url_for('goodbye', _absolute=True)
-    # OSF log out, remove current OSF session
-    osf_logout()
-    # set redirection to CAS log out (or log in if `reauth` is present)
-    if 'reauth' in request.args:
-        cas_endpoint = cas.get_login_url(redirect_url)
+    # For `?next=`:
+    #   takes priority
+    #   the url must be a valid OSF next url,
+    #   the full request url is set to CAS service url,
+    #   does not support `reauth`
+    # For `?redirect_url=`:
+    #   the url must be valid CAS service url
+    #   the redirect url is set to CAS service url.
+    #   support `reauth`
+
+    # logout/?next=<an OSF verified next url>
+    next_url = next_url or request.args.get('next', None)
```

Comment:
To be consistent with our login url: `login/?next=...`, we change the logout url from `logout/?next_url=...` to `logout/?next=...`. However, the python `variable` is still `next_url` since `next` is a reserved word.